### DISCUSSION TITLE
Repetitive CollisionDetection (batch/jobs)

### DISCUSSION THREAD

**Discussion by LukeDiederich** [pyvista/pyvista#3289] [Q&A] [answered]
Hello,

I am using PyVista at the moment to do a lot of collision detection. Basically I have a mesh (see attached file, does have a quite coarse mesh) and for every cell in the mesh I am doing one collision detection per cell.  Now for such a coarse mesh with about 40 cells that´s not the problem, but once you have a very fine mesh and a quite big geometry it does take quite a lot of time. 
For the attached example I used the .subdivide_adaptive(max_n_passes=4) method to remesh the file and then carry on with the collision detection. 

The basic order of the algorithm goes as follows:

1. reading the mesh and remeshing using mesh.subdivide_adaptive(max_n_passes=4)

```
import pyvista as pv
d = 50
Nicht_pruefbare_Zellen = []
mesh = pv.read('Kollision.stl')
mesh = mesh.subdivide_adaptive(max_n_passes=4)
mesh.plot(show_edges=True)
points = mesh.points
Normals = mesh.cell_normals
```
<img width="441" alt="image" src="https://user-images.githubusercontent.com/109736841/189187995-68211e8a-5aa3-434f-b2fe-ab3da7f4fb20.png">

--> cell number went from 44 to 11264 cells
2. creating a for loop, iterating over every cell ID from the mesh

>

 ` for cell in range(0, mesh.n_cells)`

3. for every cell I am calculating the center of mass and offsetting it for 0.01 mm in normal direction from the cell to hinder false collision resulting 

```
    Normalen = mesh.cell_normals[cell]
    Tri = mesh.cell_points(cell)
    A = Tri[0]
    B = Tri[1]
    C = Tri[2]
    S = (A + B + C) / 3

    S1 = S[0] + Normalen[0] * 0.01
    S2 = S[1] + Normalen[1] * 0.01
    S3 = S[2] + Normalen[2] * 0.01
```

4. create a disc with the center on the calculated point (S1, S2, S3) and extrude the disc in normal direction of the cell

```
    Kreis = pyvista.Disc(center=(S1, S2, S3), inner=0, outer=12,
                         normal=(Normalen[0], Normalen[1], Normalen[2]),
                         r_res=1, c_res=10)
    extrude = Kreis.extrude([Normalen[0]*d, Normalen[1]*d, Normalen[2]*d],
                            capping=True)
```

5. for this newly created mesh I analyse whether there are any intersections between the meshes "extrude" and "mesh"

```
    collision, ncol = mesh.collision(extrude, cell_tolerance=0)
    if ncol > 0:
        Nicht_pruefbare_Zellen.append(cell)
```

If there is a collision I am appending the cell to a list for it to be removed from the mesh after the analysis
For visualization I plotted them here:
<img width="483" alt="image" src="https://user-images.githubusercontent.com/109736841/189191171-b4c4f0c4-f6b2-4f70-a795-87c98d855992.png">

In the pyvista documentation (which is great by the way!) I have read, that repeated use of this method is slower than using the vtkCollisionDetectionFilter (https://docs.pyvista.org/examples/01-filter/collisions.html)
For my real analysis parts, which have a more complex geometry and much greater amounts of cells (tens to houndreds of thousands), the analysis then takes really much time.  
I also chatted with @banesullivan in short over this and we did not come to a quick solution on this topic. 

One thing I tried is to translate the before created cylinder and not create a new one for every cell:

```
        delta_x = S1 - p1_alt
        delta_y = S2 - p2_alt
        delta_z = S3 - p3_alt
        extrude.translate([delta_x, delta_y, delta_z], inplace=True)
```
where p1_alt to p3_alt are the points S1, S2 and S3 from the previous iteration. So a basic vector calculation. 
But this does not necessarily speed things up 

One final addition: 
In the vtk Filter it is also possible to 

> Get the output with the points where the contacting cells intersect (https://vtk.org/doc/nightly/html/classvtkCollisionDetectionFilter.html#details)

which is also a feature which would be very nice if it could be implemented in this code as well. 

So to sum up: 
I need a quicker approach to do repeated collision analysis with, where my input are two meshes and the output is whether the meshes intersect (basically number of collisions) as well as the points where the cells intersect. Currently I am referecing the cells which intersect and then calculate their center of mass and then calculate the distance between both center of masses to have a basic approximation. 

Any help is greatly appreciated

Thanks in advance
Luke 

[Kollision.zip](https://github.com/pyvista/pyvista/files/9528965/Kollision.zip)

**Comment by banesullivan** [comment]
Thank you fo summarizing this, @LukeDiederich! We chatted about this on a video call and I'm working asynchronously to see how I can help optimize this

**Comment by LukeDiederich** [comment]
One thing I would like to add: 

What I found is that the finer the mesh "mesh" in this example is, the more time the analysis takes, which does make sense.

A possible solution would be that the ouput from vtk (Get the output with the points where the contacting cells intersect) is somehow added to the existing collision detection in pyvista or something in that direction. 
This would enable the analysis to run with a quite coarse mesh (using the original mesh instead of the remeshed mesh) and would therefore be faster.
As a recap: I need the finer mesh, because for distance computing I used the center of mass from the collision cells. For a very coarse mesh this method is quite inaccurate.

This was something that came into my mind. Hope it makes any sense and helps anyhow.

Code would then be edited to look like this:

`coarse_mesh = pv.read('Kollision.stl')`

Code in between remains the same!

```
collision, ncol = coarse_mesh.collision(extrude, cell_tolerance=0)
    if ncol > 0:
        Nicht_pruefbare_Zellen.append(cell)
```

This small adjustment improves the time of the collision detection drastically (dependent on how much finer the mesh is), but reduces the accuracy of the distance computation significantly.

**Comment by banesullivan** [comment]
Following up.... I'm having a lot of trouble speeding this up

**Reply by LukeDiederich** [reply]
If I can be of any sort of help please just let me know, I´ll be glad to at least try to help

If at the moment this is a "limitation" of PyVista I can live with that, at least we tried to look into it, which I´m more than glad about.

**Comment by banesullivan** [comment, marked as answer]
here is a first pass at parallelizing the code which shows some promising speedup

- Run sequentially: has been running for several minutes (not done yet)
- Run parallel: ~74 seconds


```py
import multiprocessing
import time

import numpy as np
import pyvista as pv
import vtk

pv.set_error_output_file("errors.txt")
vtk.vtkObject.GlobalWarningDisplayOff()

og = pv.read('Kollision.stl')
mesh = og.subdivide_adaptive(max_n_passes=4)


class Collision:
    def __init__(self, mesh, d=50):
        self.mesh = mesh
        self.d = d

    def __call__(self, cell):
        Normalen = self.mesh.cell_normals[cell]
        Tri = self.mesh.cell_points(cell)
        A = Tri[0]
        B = Tri[1]
        C = Tri[2]
        S = (A + B + C) / 3

        S1 = S[0] + Normalen[0] * 0.01
        S2 = S[1] + Normalen[1] * 0.01
        S3 = S[2] + Normalen[2] * 0.01

        Kreis = pv.Disc(center=(S1, S2, S3), inner=0, outer=12,
                            normal=(Normalen[0], Normalen[1], Normalen[2]),
                            r_res=1, c_res=10)
        extrude = Kreis.extrude([Normalen[0]*self.d, Normalen[1]*self.d, Normalen[2]*self.d],
                                capping=True)


        collision, ncol = self.mesh.collision(extrude, cell_tolerance=0)

        if ncol > 0:
            return cell
        return -1


if __name__ == '__main__':
    start_time = time.time()

    collision = Collision(mesh)

    with multiprocessing.Pool() as p:
        # Execution time in seconds: 74.34431791305542
        results = p.map(collision, range(mesh.n_cells))

    # for cell in range(mesh.n_cells):
    #     # Execution time in seconds: ?
    #     collision(cell)

    execution_time = (time.time() - start_time)
    print('Execution time in seconds: ' + str(execution_time))
```

**Reply by LukeDiederich** [reply]
Thanks @banesullivan!
We actually chatted over this in a quick call and together came to the conclusion that it is the easiest solution to try to parallelize the process (e.g. multiprocessing)
I´ll try to implement it or come up with another solution, we´ll see
